stablehlo/reference/Api.cpp

```diff
@@ -137,6 +139,39 @@ LogicalResult validateEntrySignature(func::FuncOp func,
   return success();
 }
 
+// Specializes the shapes of arguments in function 'func' based on runtime input
+// values. If all argument types already have static shapes, this function does
+// nothing. Otherwise, it constructs a pipeline of MLIR passes to refine
+// argument shapes using the provided `inputs`.
+//
+// Args:
+//   module: The MLIR module containing the function.
+//   func: The function whose argument shapes need specialization.
+//   inputs: The runtime input values.
+//
+// Returns:
+//   A `LogicalResult` indicating success or failure of the shape
+//   refinement pipeline.
+LogicalResult refinePolymorphicModule(ModuleOp module, func::FuncOp func,
+                                      ArrayRef<InterpreterValue> inputs) {
+  if (llvm::all_of(func.getArgumentTypes(), [](Type type) {
+        return llvm::cast<ShapedType>(type).hasStaticShape();
+      })) {
+    return success();
+  }
+
+  SmallVector<Type> refinedTypes = llvm::to_vector(llvm::map_range(
+      inputs, [](InterpreterValue input) { return input.getType(); }));
+
+  PassManager pm(module.getContext());
+  stablehlo::createStablehloRefineArgumentsAndShapesPipeline(pm, refinedTypes);
```

Comment:
What happens if the number of function arguments does not match the number of shapes? Consider adding a test case for that. I think we should probably validate that before applying any passes.

Reply:
The first pass in the pipeline has the check for that https://github.com/openxla/stablehlo/blob/e2201e2af69c2b0d32370cfe0deafe2681c7b3e9/stablehlo/transforms/StablehloRefineArguments.cpp#L93 and we have tests available as well. Please let me know if that would suffice. 
 
I note that [validateEntrySignature](https://github.com/openxla/stablehlo/blob/e2201e2af69c2b0d32370cfe0deafe2681c7b3e9/stablehlo/reference/Api.cpp#L120) in APIs.cpp is already doing that check, but I reorder it to be after the the shape specialization mainly because that function deals with static shapes only and with that it is very easy to match types using `a == b` as opposed to do compatibility check ignoring dynamic dimensions for all supported interpreter `Tensor` types (Tuples and Tenor).

Reply:
That sounds reasonable. I'm just wondering what the error message will look like and if it will be clear to the user what the issue is?

LGTM about `validateEntrySignature`. Earlier on I think some comment mentioned that check ran first, which I think couldn't have worked because the types would have differed. Doing the check after removing dynamism makes sense to me. Thanks!

Reply:
Fair point. Here is an example.
https://github.com/openxla/stablehlo/blob/e2201e2af69c2b0d32370cfe0deafe2681c7b3e9/stablehlo/tests/stablehlo_refine_parameters.mlir#L39

Reply:
This looks fine, I just wonder if it will make sense to someone who's just using the interpreter and has no knowledge of what's going on under the hood. IMO it would be preferable to do this check before running the pipeline as that would allow us to produce a more helpful error message.

Up to you though, I think what we have now is good enough.

Reply:
Sure. Lets check is in as in and we can refine later.